service_provider.go

```diff
@@ -903,44 +906,9 @@ func (sp *ServiceProvider) parseResponse(responseEl *etree.Element, possibleRequ
 	}
 
 	// validate request attributes
-	{
-		var response Response
-		if err := unmarshalElement(responseEl, &response); err != nil {
-			return nil, fmt.Errorf("cannot unmarshal response: %v", err)
-		}
-
-		if !sp.SkipDestinationCheck || response.Destination != "" {
-			// If the response is *not* signed, the Destination may be omitted.
-			if responseHasSignature || response.Destination != "" {
-				if response.Destination != sp.AcsURL.String() {
-					return nil, fmt.Errorf("`Destination` does not match AcsURL (expected %q, actual %q)", sp.AcsURL.String(), response.Destination)
-				}
-			}
-		}
-
-		requestIDvalid := false
-		if sp.AllowIDPInitiated {
-			requestIDvalid = true
-		} else {
-			for _, possibleRequestID := range possibleRequestIDs {
-				if response.InResponseTo == possibleRequestID {
-					requestIDvalid = true
-				}
-			}
-		}
-		if !requestIDvalid {
-			return nil, fmt.Errorf("`InResponseTo` does not match any of the possible request IDs (expected %v)", possibleRequestIDs)
-		}
-
-		if response.IssueInstant.Add(MaxIssueDelay).Before(now) {
-			return nil, fmt.Errorf("response IssueInstant expired at %s", response.IssueInstant.Add(MaxIssueDelay))
-		}
-		if !sp.SkipIssuerCheck && response.Issuer != nil && response.Issuer.Value != sp.IDPMetadata.EntityID {
-			return nil, fmt.Errorf("response Issuer does not match the IDP metadata (expected %q)", sp.IDPMetadata.EntityID)
-		}
-		if response.Status.StatusCode.Value != StatusSuccess {
-			return nil, ErrBadStatus{Status: response.Status.StatusCode.Value}
-		}
+	validateErr := sp.validateRequestAttributes(responseEl, possibleRequestIDs, now, responseHasSignature)
```

Comment:
Moved validate request attributes into a method to fix cyclomatic complexity lint issue